### PR TITLE
Add Fake.ResetToInitialConfiguration

### LINKED ISSUE #1839: Discuss alternatives to removing ClearConfiguration
In #1764, @PiotrKlecha raised a concern about how `ClearConfiguration` removes the strictness from a [strict Fake](https://fakeiteasy.readthedocs.io/en/stable/strict-fakes/), and this was unexpected, since the strictness is (usually) applied at creation time, and so it feels like it's an intrinsic aspect of the Fake, rather than some additional configuration that's added on after.

In that case, it turned out that there was a way to reformulate @PiotrKlecha's tests so ClearConfiguration wasn't needed, but the bell couldn't be unrung. @thomaslevesque and I discussed how ClearConfiguration worked and how it's very difficult to predict what it would do from the outside. Complicating factors that we identified

* configuration to make the Fake strict, to have it wrap another object, or to have it call base methods are typically applied at creation time, and may not be what the client has in mind when clearing configuration
* configuration to make a Fake act strict or to have it call a base method can be applied after creation time, although it's less convenient. Therefore retaining particular kinds of rules may not be easy.
* [implicit creation options](https://fakeiteasy.readthedocs.io/en/stable/implicit-creation-options/) further confound the picture. Since they're not in front of the user, they may be relied on without them knowing about it, so removing them from the configuration might be a surprise (conversely, _not_ removing them might be a surprise).

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #1839

### CODE REVIEW

File: tests/FakeItEasy.Specs/FakeSpecs.cs [hunk 1]
```diff
@@ -141,6 +149,125 @@ public static void ClearConfiguration(IFoo fake, bool configuredBehaviorWasUsed)
...
+        }
+
+        [Scenario]
+        public static void ResetToInitialConfigurationKeepsAutoPropertyUsedInConstructor(SetsVirtualPropertyInConstructor fake, string? result)
+        {
+            "Given a Fake whose constructor sets the value of a read/write property"
+                .x(() => fake = A.Fake<SetsVirtualPropertyInConstructor>());
+
+            "When I reset the configuration to its initial state"
+                .x(() => Fake.ResetToInitialConfiguration(fake));
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply tests/FakeItEasy.Specs/FakeSpecs.cs:262]: (I guess that would also apply to the other tests)

**Review by thomaslevesque — COMMENTED** [review]
Thanks @blairconrad. Looks great to me!
Just one question and a suggestion for the specs

File: src/FakeItEasy/Core/FakeManager.cs [hunk 2]
```diff
@@ -29,6 +29,8 @@ public partial class FakeManager : IFakeCallProcessor
 
         private readonly LinkedList<CallRuleMetadata> allUserRules;
 
+        private LinkedList<CallRuleMetadata>? initialUserRules;
```
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/Core/FakeManager.cs:32]: Lazy copying.

**Comment by blairconrad** [comment]
Closing for now. The implementation will not be correct, and the tests, even after updating as @thomaslevesque suggests, will be wildly inadequate, as he pointed out in side chat. See #1839 for an accounting of the discussion.

**Pull request closed without merging**